opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDateType.java

```diff
@@ -198,38 +218,71 @@ public List<DateFormatter> getTimeNamedFormatters() {
     return getFormatList().stream()
         .filter(formatString -> {
           FormatNames namedFormat = FormatNames.forName(formatString);
-          return SUPPORTED_NAMED_TIME_FORMATS.contains(namedFormat);
+          return namedFormat == null ? false : SUPPORTED_NAMED_TIME_FORMATS.contains(namedFormat);
         })
         .map(DateFormatter::forPattern).collect(Collectors.toList());
   }
 
-  private ExprCoreType getExprTypeFromFormatString(String formatString) {
-    if (formatString.isEmpty()) {
-      // FOLLOW-UP: check the default formatter - and set it here instead
-      // of assuming that the default is always a timestamp
-      return TIMESTAMP;
+  private ExprCoreType getExprTypeFromCustomFormatString(String formatString) {
+    List<String> formats = Arrays.asList(formatString.split("\\|\\|"));
+    boolean isDate = false;
+    boolean isTime = false;
+
+    try {
+      isDate = formats.stream()
+          .anyMatch(format -> DateTimeFormatter.ofPattern(format).toString().toLowerCase().contains("year"));
```

Comment:
What if format contains only [`year`](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-date-format.html#:~:text=year%20or%20strict_year-,A%20formatter%20for%20a%20four%20digit%20year,-%3A%20yyyy.)?

Reply:
If the format only includes "yyyy", then Formatter would still contain "year" which will make `isDate` true. If only month is in the format, then Formatter string still contains "year". 

In terms of time, an hour is always required in order to load the data in to OpenSearch. So I have changed the check to "hour" instead of "minute".

Reply:
Interesting observation. 
We should have a test for each of these cases:  a custom format with just "mm" and custom format with just "MM" to test the two 'm'-only cases. 
What's the expected result in these cases?   A failure to format, or datetime object?

Reply:
We should test that in runtime to understand whether it is date or time or datetime. We should test it
* for every `date` column with custom formats
* for every index involved in a query
* for every query

Unfortunately, we don't cache that, but it is possible...

I see two ways to test it, but I dislike both due to performance reasons:
1. Generate a sample input (like "11:11:11 1111-11-11") for each format and parse it. Then check which fields are discovered.
2. Convert `LocalDateTime.now()` by each format and parse it back. Then check which fields are discovered. This option is more reliable.